### ISSUE TITLE
Populate integration with cache

### ISSUE THREAD

**Issue by yl2chen** [Automattic/mongoose#2068]
Correct me if I am wrong but I understood mongoose'e implementation for populate as a a direct db query everytime.

It would be super nice for mongoose to support caching for objects that are populated as a configurable option (let user incorporate their own caching implementation), because currently what I would do is get the object and manually populate any reference from the cache first (then hit db if not in cache).  I am not so comfortable with using the populate feature, although it is super convenient, because the extra read it incurs each time.

tldr: Does mongoose currently have caching layer support for populate feature?  Or any suggestions?

**Comment by DominicBoettger** [comment]
Back to the question: Do you know any plugin or is it planned to replace this slow solution?

**Comment by vkarpov15** [comment]
Not planned atm. With query middleware in 4.x you could create a plugin that caches query results at the schema level, which would solve this since `populate()` is just syntactic sugar for `MyModel.find()`. You can just cache the whole result of `MyModel.find().populate()` in your application though.

**Comment by DominicBoettger** [comment]
Thanks for your reply. If i add a pre find event. How can i provide the result and skip the original find? I did not find a solution for this use case. The only way to skip the find seems to add a error to next...

**Comment by vkarpov15** [comment]
Yeah the only way that I know of right now would be to pass an error to `next()` to prevent the actual find, and then check the type of the error. That will create a cache for *all* finds on that model though.

An alternative approach would be to overwrite `.populate()` with a wrapper that checks the cache.

**Comment by evilive3000** [comment]
Did anything change from that time? are there more convenient ways for setting cached data on population? 
I have set of `tag` documents and it's a small set, so I want to fetch it from db and put into memory once.

**Comment by DominicBoettger** [comment]
Some time ago i did a implementation for a problem like that in the database layer of molecuel.... May this is helpful for you... https://github.com/molecuel/mlcl_database_cache/blob/master/index.ts
